bfloat16: enable basic math function

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35172 bfloat16: enable basic math function**

Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

XiaobingSuper committed:
bfloat16: enable basic math function

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/cpu/vml.h:
```diff
@@ -39,13 +39,10 @@
-  x = std::op(x);                                             \
-  _mm256_zeroall();
+// Avoid AVX/SSE transition overhead
+#define DL_RUNTIME_BUG _mm256_zeroall();
```
just out of curiosity, why did this change?

XiaobingSuper replied:
for glic=2.23, **x = std::op(x);** x should have a volatile type, but for BFloat16, such as **c10::BFloat16 acos(c10::BFloat16 a) { return std::acos(float(a));}**, the parameter is not a volatile type.

ngimel replied:
But you are changing it here not only for BFLoat16, but for all types?

cpuhrsch replied:
Please don't touch this. If you need something for a one-off type, please specialize the code.

XiaobingSuper replied:
@cpuhrsch , changed as your suggestion. Thanks!

ngimel commented on test/test_torch.py:
```diff
@@ -16598,33 +16598,45 @@ def inner(self, device, dtype):
-    ('cos', '', _small_3d, lambda t, d: [], 1e-3, 1e-5, 1e-5, _float_types),
+    ('log', '', _small_3d, lambda t, d: [], 1e-2, 1e-2, 1e-5, _float_types2),
+    ('log', 'no_zero', lambda t, d: _small_3d(t, d, has_zeros=False),
+        lambda t, d: [], 1e-2, 1e-1, 1e-5, _float_types2, [torch.bfloat16]),
```
nit: it may make sense to have a variable, say, _low_prec_types equal to [torch.bfloat16], similar to how all other type lists are specified.

XiaobingSuper replied:
For some ops, the ouput can be a large number, the absolute error can be a big number, relative error will be a good choice, I will try it at next step.

ngimel commented on test/test_torch.py:
```diff
@@ -16598,33 +16598,45 @@ def inner(self, device, dtype):
-    ('atan', '', _small_3d, lambda t, d: [], 1e-3, 1e-5, 1e-5, _float_types),
-    ('cos', '', _small_3d, lambda t, d: [], 1e-3, 1e-5, 1e-5, _float_types),
+    ('log', '', _small_3d, lambda t, d: [], 1e-2, 1e-2, 1e-5, _float_types2),
+    ('log', 'no_zero', lambda t, d: _small_3d(t, d, has_zeros=False),
```
similar to reciprocal, why do you need no_zero here?

ngimel reviewed (commented):
Looks good overall, I'm not qualified to review Sleef/cpu intrinsics part, but tests should cover it.

ngimel commented:
Why did you close this?

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

dlibenzi commented:
@ailzhang Seems like the new OP registration code changed and we got broken.

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

ngimel commented:
@XiaobingSuper we have internal build errors on this pr 
```
Summary: 
caffe2/c10/util/BFloat16-inl.h:285:45: error: all paths through this function will call itself [-Werror,-Winfinite-recursion]
inline c10::BFloat16 rsqrt(c10::BFloat16 a) { return std::rsqrt(float(a));}
                                            ^
caffe2/c10/util/BFloat16-inl.h:285:45: error: all paths through this function will call itself [-Werror,-Winfinite-recursion]
inline c10::BFloat16 rsqrt(c10::BFloat16 a) { return std::rsqrt(float(a));}
                                            ^
caffe2/c10/util/BFloat16-inl.h:285:45: error: all paths through this function will call itself [-Werror,-Winfinite-recursion]
inline c10::BFloat16 rsqrt(c10::BFloat16 a) { return std::rsqrt(float(a));}
                                            ^
```
can you take a look at what might be causing it?

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper commented:
@ngimel , the is caused by **std** have not **rsqrt** function, it only has this support in cuda.

ngimel commented:
Now there are similar errors for erf, erfc, expm1 and lgamma:
```
Summary: 
xplat/caffe2/c10/util/BFloat16-inl.h:267:43: error: all paths through this function will call itself [-Werror,-Winfinite-recursion]
inline c10::BFloat16 erf(c10::BFloat16 a) { return std::erf(float(a));}
                                          ^
xplat/caffe2/c10/util/BFloat16-inl.h:268:44: error: all paths through this function will call itself [-Werror,-Winfinite-recursion]
inline c10::BFloat16 erfc(c10::BFloat16 a) { return std::erfc(float(a));}
                                           ^
xplat/caffe2/c10/util/BFloat16-inl.h:270:45: error: all paths through this function will call itself [-Werror,-Winfinite-recursion]
inline c10::BFloat16 expm1(c10::BFloat16 a) { return std::expm1(float(a));}
                                            ^
xplat/caffe2/c10/util/BFloat16-inl.h:273:45: error: all paths through this function will call itself [-Werror,-Winfinite-recursion]
inline c10::BFloat16 log1p(c10::B ...(681595 bytes)... e-recursion]
inline c10::BFloat16 lgamma(c10::BFloat16 a) { return std::lgamma(float(a));}
                                             ^
```

XiaobingSuper commented:
@ngimel , this failed case happend on mobile platform? it will be help to sovle this problem to see details build log, such as which C++ version used?

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper commented:
@ngimel , could you re-land it to see whether the former failed case is passed now?

ngimel commented:
I've reimported it, will let you know. If mobile builds still fail, you can try guarding bfloat16 code so that it's not compiled in mobile builds.

ngimel commented:
It still fails with the same errors. You can also try submitting this PR to a branch ending with `ci_all` to run all OSS CI tests on it, maybe it'll be reproduced somewhere. If not, try guarding on mobile builds.

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]

XiaobingSuper commented:
@ngimel, I think it works now, the reason is that some functions not defined in **std** namespace for old Android versions, see https://github.com/pytorch/pytorch/blob/b92f8d9b7e2e8543e2808c1d05c09e57c19ce5d9/c10/util/math_compat.h#L5-L7

XiaobingSuper committed:
Update on "bfloat16: enable basic math function"


Differential Revision: [D20721146](https://our.internmc.facebook.com/intern/diff/D20721146)

[ghstack-poisoned]
